funsor/torch/provenance.py

```diff
@@ -0,0 +1,48 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+
+class ProvenanceTensor(torch.Tensor):
+    def __new__(cls, data, provenance=frozenset(), **kwargs):
+        if not provenance:
+            return data
+        instance = torch.Tensor.__new__(cls)
+        instance.__init__(data, provenance)
+        return instance
+
+    def __init__(self, data, provenance=frozenset()):
+        assert isinstance(provenance, frozenset)
+        if isinstance(data, ProvenanceTensor):
+            provenance |= data._provenance
+            data = data._t
+        self._t = data
+        self._provenance = provenance
+
+    def __repr__(self):
+        return "Provenance:\n{}\nTensor:\n{}".format(self._provenance, self._t)
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
```

Comment:
@ordabayevy now that you've had a chance to play around with `__torch_function__`, I'm curious about whether you think we should add a `Funsor.__torch_function__` method and attempt to use it in Pyro. I opened #546 to discuss.